File: javascript/tests/test_traversal.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, mkdirSync, writeFileSync, chmodSync, symlinkSync, unlinkSync, rmdirSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import { walk } from '../src/traversal.js';

function mktmp() {
  return mkdtempSync(join(tmpdir(), 'scanner-test-'));
}

test('returns files in directory', () => {
  const d = mktmp();
  writeFileSync(join(d, 'app.py'), '');
  writeFileSync(join(d, 'main.js'), '');
  const { files, skipped } = walk(d);
  assert.equal(files.length, 2);
  assert.equal(skipped, 0);
});

test('ignores node_modules', () => {
  const d = mktmp();
  mkdirSync(join(d, 'node_modules'));
  writeFileSync(join(d, 'node_modules', 'pkg.js'), '');
  writeFileSync(join(d, 'app.py'), '');
  const { files } = walk(d);
  assert.equal(files.length, 1);
  assert.ok(files[0].endsWith('app.py'));
});

test('ignores all configured dirs', () => {
  const ignoreDirs = ['.git', 'node_modules', '__pycache__', 'venv', '.env', 'dist', 'build'];
  const d = mktmp();
  for (const name of ignoreDirs) {
    mkdirSync(join(d, name));
    writeFileSync(join(d, name, 'file.js'), '');
  }
  writeFileSync(join(d, 'keep.py'), '');
  const { files } = walk(d);
  assert.equal(files.length, 1);
});

test('does not follow symlinks', () => {
  const d = mktmp();
  const real = join(d, 'real.py');
  writeFileSync(real, '');
  symlinkSync(real, join(d, 'link.py'));
  const { files } = walk(d);
  assert.equal(files.length, 1);
  assert.ok(files[0].endsWith('real.py'));
});

test('walks subdirectories', () => {
  const d = mktmp();
  mkdirSync(join(d, 'sub'));
  writeFileSync(join(d, 'a.py'), '');
  writeFileSync(join(d, 'sub', 'b.py'), '');
  const { files } = walk(d);
  assert.equal(files.length, 2);
});

File: python/tests/test_traversal.py
import os
import tempfile
import pytest
from scanner.traversal import walk


def test_files_returned():
    with tempfile.TemporaryDirectory() as d:
        open(os.path.join(d, 'app.py'), 'w').close()
        open(os.path.join(d, 'main.js'), 'w').close()
        files, skipped = walk(d)
        assert len(files) == 2
        assert skipped == 0


def test_ignore_dirs():
    with tempfile.TemporaryDirectory() as d:
        nm = os.path.join(d, 'node_modules')
        os.makedirs(nm)
        open(os.path.join(nm, 'pkg.js'), 'w').close()
        open(os.path.join(d, 'app.py'), 'w').close()
        files, skipped = walk(d)
        assert len(files) == 1
        assert files[0].endswith('app.py')


def test_all_ignore_dirs():
    ignore_names = ['.git', 'node_modules', '__pycache__', 'venv', '.env', 'dist', 'build']
    with tempfile.TemporaryDirectory() as d:
        for name in ignore_names:
            subdir = os.path.join(d, name)
            os.makedirs(subdir)
            open(os.path.join(subdir, 'file.py'), 'w').close()
        open(os.path.join(d, 'keep.py'), 'w').close()
        files, _ = walk(d)
        assert len(files) == 1


def test_symlinks_not_followed():
    with tempfile.TemporaryDirectory() as d:
        real = os.path.join(d, 'real.py')
        open(real, 'w').close()
        link = os.path.join(d, 'link.py')
        os.symlink(real, link)
        files, skipped = walk(d)
        assert len(files) == 1
        assert os.path.basename(files[0]) == 'real.py'


def test_recursive_walk():
    with tempfile.TemporaryDirectory() as d:
        sub = os.path.join(d, 'sub')
        os.makedirs(sub)
        open(os.path.join(d, 'a.py'), 'w').close()
        open(os.path.join(sub, 'b.py'), 'w').close()
        files, skipped = walk(d)
        assert len(files) == 2
        assert skipped == 0


def test_unreadable_dir_counted():
    with tempfile.TemporaryDirectory() as d:
        sub = os.path.join(d, 'secret')
        os.makedirs(sub)
        open(os.path.join(sub, 'file.py'), 'w').close()
        os.chmod(sub, 0o000)
        try:
            files, skipped = walk(d)
            assert skipped >= 1
        finally:
            os.chmod(sub, 0o755)
